<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>SC-MEB CRC</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">SC-MEB CRC</h1>



<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(<span class="st">&quot;SC.MEB&quot;</span>)</span></code></pre></div>
<div id="fit-sc-meb-using-real-data-crc" class="section level2">
<h2>Fit SC-MEB using real data CRC</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>file =<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;CRC3.rds&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;SC.MEB&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a>CRC =<span class="st"> </span><span class="kw">readRDS</span>(file)</span></code></pre></div>
<div id="pre-processing-data" class="section level3">
<h3>Pre-processing data</h3>
<p>SC-MEB requires minimal data pre-processing, but we provide a helper function to automate it.</p>
<p>spatialPreprocess() log-normalizes the count matrix and performs PCA on the top n.HVGs highly variable genes, keeping the top n.PCs principal components. Additionally, the spatial sequencing platform is added as metadata in the SingleCellExperiment for downstream analyses. If you do not wish to rerun PCA, running spatialPreprocess() with the flag skip.PCA=TRUE will only add the metadata SC-MEB requires.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw">set.seed</span>(<span class="dv">114</span>)</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="kw">library</span>(scuttle)</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="kw">library</span>(scran)</span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="kw">library</span>(scater)</span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="kw">library</span>(BiocSingular)</span>
<span id="cb3-6"><a href="#cb3-6"></a>CRC &lt;-<span class="st"> </span><span class="kw">spatialPreprocess</span>(CRC, <span class="dt">platform=</span><span class="st">&quot;Visium&quot;</span>)</span></code></pre></div>
<p>Here, we set the basic paramters for our function <code>SC.MEB</code></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>platform =<span class="st"> &quot;Visium&quot;</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>beta_grid =<span class="st"> </span><span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">4</span>,<span class="fl">0.2</span>)</span>
<span id="cb4-3"><a href="#cb4-3"></a>K_set=<span class="st"> </span><span class="dv">2</span><span class="op">:</span><span class="dv">10</span></span>
<span id="cb4-4"><a href="#cb4-4"></a>parallel=<span class="ot">TRUE</span></span>
<span id="cb4-5"><a href="#cb4-5"></a>num_core =<span class="st"> </span><span class="dv">3</span></span>
<span id="cb4-6"><a href="#cb4-6"></a>PX =<span class="st"> </span><span class="ot">TRUE</span></span>
<span id="cb4-7"><a href="#cb4-7"></a>maxIter_ICM =<span class="st"> </span><span class="dv">10</span></span>
<span id="cb4-8"><a href="#cb4-8"></a>maxIter =<span class="st"> </span><span class="dv">50</span></span></code></pre></div>
</div>
<div id="fitting-the-sc-meb" class="section level3">
<h3>Fitting the SC-MEB</h3>
<div id="calculating-the-neighborhood" class="section level4">
<h4>Calculating the neighborhood</h4>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">library</span>(SingleCellExperiment)</span>
<span id="cb5-2"><a href="#cb5-2"></a>Adj_sp  &lt;-<span class="st"> </span><span class="kw">find_neighbors2</span>(CRC, <span class="dt">platform =</span> <span class="st">&quot;Visium&quot;</span>)</span>
<span id="cb5-3"><a href="#cb5-3"></a>Adj_sp[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span></code></pre></div>
</div>
<div id="run-the-sc-meb-in-parallel" class="section level4">
<h4>Run the SC-MEB in parallel</h4>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>y =<span class="st"> </span><span class="kw">reducedDim</span>(CRC, <span class="st">&quot;PCA&quot;</span>)[,<span class="dv">1</span><span class="op">:</span><span class="dv">15</span>]</span>
<span id="cb6-2"><a href="#cb6-2"></a>fit =<span class="st"> </span><span class="kw">SC.MEB</span>(y, Adj_sp, <span class="dt">beta_grid =</span> beta_grid, <span class="dt">K_set=</span> K_set, <span class="dt">parallel=</span>parallel, <span class="dt">num_core =</span> num_core, <span class="dt">PX =</span> PX, <span class="dt">maxIter_ICM=</span>maxIter_ICM, <span class="dt">maxIter=</span>maxIter)</span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="kw">str</span>(fit[,<span class="dv">1</span>])</span></code></pre></div>
</div>
</div>
<div id="clustering" class="section level3">
<h3>Clustering</h3>
<div id="selecting-the-number-of-clusters-using-bic" class="section level4">
<h4>Selecting the number of clusters using BIC</h4>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">selectKPlot</span>(fit, <span class="dt">K_set =</span> K_set, <span class="dt">criterion =</span> <span class="st">&quot;BIC&quot;</span>)</span></code></pre></div>
</div>
<div id="selecting-the-number-of-clusters-using-modified-bic" class="section level4">
<h4>Selecting the number of clusters using Modified BIC</h4>
<p>Here we briefly explain how to choose the parameter c in the modified BIC. In general, For the ST or Visium dataset, it often ranges from 0.4 to 1 while for the MERFISH dataset with large number of cells, it often becomes larger, for example 10,20. Most importantly, SC-MEB is fast, scaling well in terms of sample size, which allow the user to tune the c based on their prior knowledge about the tissues or cells.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw">selectKPlot</span>(fit, <span class="dt">K_set =</span> K_set, <span class="dt">criterion =</span> <span class="st">&quot;MBIC&quot;</span>)</span></code></pre></div>
</div>
<div id="visualizing-spatial-clusters" class="section level4">
<h4>Visualizing spatial clusters</h4>
<p>We can plot the cluster assignments over the spatial locations of the spots with ClusterPlot().</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>out =<span class="st"> </span><span class="kw">selectK</span>(fit, <span class="dt">K_set =</span> K_set, <span class="dt">criterion =</span> <span class="st">&quot;BIC&quot;</span>)</span>
<span id="cb9-2"><a href="#cb9-2"></a>pos =<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">cbind</span>(<span class="kw">colData</span>(CRC)[,<span class="kw">c</span>(<span class="dv">4</span>)],<span class="dv">20000</span><span class="op">-</span><span class="kw">colData</span>(CRC)[,<span class="kw">c</span>(<span class="dv">3</span>)]), <span class="dv">2988</span>, <span class="dv">2</span>)</span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="kw">ClusterPlot</span>(out, pos, <span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">shape =</span> <span class="dv">16</span>)</span></code></pre></div>
<p>or</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>out =<span class="st"> </span><span class="kw">selectK</span>(fit, <span class="dt">K_set =</span> K_set, <span class="dt">criterion =</span> <span class="st">&quot;MBIC&quot;</span>)</span>
<span id="cb10-2"><a href="#cb10-2"></a>pos =<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">cbind</span>(<span class="kw">colData</span>(CRC)[,<span class="kw">c</span>(<span class="dv">4</span>)],<span class="dv">20000</span><span class="op">-</span><span class="kw">colData</span>(CRC)[,<span class="kw">c</span>(<span class="dv">3</span>)]), <span class="dv">2988</span>, <span class="dv">2</span>)</span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="kw">ClusterPlot</span>(out, pos, <span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">shape =</span> <span class="dv">16</span>)</span></code></pre></div>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
